apple/internal/binary_support.bzl

```diff
@@ -240,6 +244,15 @@ def _create_binary(
     linkopts = linkopts + collections.before_each("-rpath", rule_descriptor.rpaths)
     linkopts = linkopts + rule_descriptor.extra_linkopts
 
+    # Special case for `ios_extension` until it is migrated to the Starlark linking
+    # API. Note that we use `kwargs.get` instead of `pop` so that the attribute
+    # remains in the dictionary to be passed to the bundling rule; this ensures that
+    # analysis fails if the user tries to pass the attribute to a rule that doesn't
+    # support it.
+    provides_main = kwargs.get("provides_main", False)
```

Comment:
Can’t this be derived from `sdk_frameworks` (if it contains SwiftUI or not)? (I’m not sure if non-SwiftUI app extension with a `main` entry point is allowed or not)

Reply:
`swift_library` doesn't support `sdk_frameworks` because we still use auto-linking for Swift modules imported from the SDK, so there's no requirement to list is separately. The `ios_extension` rule might _accidentally_ support it because of the way we pass arguments from the macro down to the underlying rule, but that's not documented and will go away when we migrate the iOS rules away from `apple_binary` (as we've already done for the other platforms).

Even if we had that information reliably, other kinds of extensions that still use the legacy `_NSExtensionMain` entry point could use SwiftUI for their interfaces if they wished (via `UIHostingController`), so just linking to SwiftUI isn't enough to make this decision.

An alternative I've considered is adding an implicit dependency that adds its own custom entry point that is named *neither* `_main` nor `_NSExtensionMain`, and that function would be a small trampoline that checks for the presence of `_main` in the binary and calls it if present, otherwise it calls `_NSExtensionMain`. That's actually what `_WKExtensionMain` does for watchOS 7 apps (I'm not sure why Apple didn't modify `_NSExtensionMain` to have the same behavior), but I'm always wary about diverging in weird ways from how such a binary would be traditionally built, and I would want to wait for Xcode 12/iOS 14/macOS 11 to go GM before trying it out to make sure that it can be tested in an actual production environment and App Store submission.

Reply:
Thanks for the explaination. Xcode 12 beta does this automatically, which made me wonder which build setting it looks for to make that decision.

Reply:
In Xcode, you explicit say which type of extension you're creating when you add a new target—in this case a Widget Extension—so it knows exactly what special behavior it needs to enable for each extension.

Ultimately though, since it's otherwise just an "app extension" that's only distinguished by its entry point, it's easier to support in Bazel using this attribute than by trying to create a wholly separate concept for it; especially if Apple starts supporting `@main` for other kinds of extensions in the future.

Reply:
BTW, it seems the only thing Xcode sees is the values in Info.plist:
```xml
<key>NSExtension</key>
<dict>
	<key>NSExtensionPointIdentifier</key>
	<string>com.apple.widgetkit-extension</string>
</dict>
```

Otherwise the setup is exactly the same as any other app extension.

Reply:
This will not work as you expect. Widget extension require the _NSExtensionMain entry point.